### DISCUSSION TITLE
Applying a LUT to an image?

### DISCUSSION THREAD

**Discussion by hadmut** [ImageMagick/ImageMagick6#348] [General]
Hi,

I am looking for a way to apply a LUT (preferrably in .cube format) to a single image from command line/scripts, mainly to visualize the effect of a LUT table. Unfortunately, image magick's documentation is not exactly talkative about this, 

https://imagemagick.org/script/command-line-options.php#clut
https://usage.imagemagick.org/color_mods/#clut

I would have expect that something like

convert in.jpg -clut RED_Achromic_Rec2020_N-Log_to_Rec709_BT1886.cube out.png

But under Ubuntu 24.04 (still with ImageMagick 6.9.12-98) there's an error message:

% convert in.jpg -clut RED_Achromic_Rec2020_N-Log_to_Rec709_BT1886.cube out.png
convert-im6.q16: image sequence is required `-clut' @ error/mogrify.c/MogrifyImageList/7921.
convert-im6.q16: no decode delegate for this image format `CUBE' @ error/constitute.c/ReadImage/581.
convert-im6.q16: no images defined `out.png' @ error/convert.c/ConvertImageCommand/3234.

Since there (at least in the 7.x) versions is a coders/cube.c , I would have expected to find cube in the output of 

convert -list format 

but it isn't listed. 

So my questions are:

1. What's the correct way to apply a .cube LUT on a general image?
2. Does the fact that .cube is not listed in output of convert -list format indicate that the coder for cubes is not compiled into the ubuntu version?

best regards

**Comment by hadmut** [comment]
Alpine linux uses ImageMagick 7.1.1 and lists CUBE in convert -list format , but still gives an error message:


# convert in.jpg -clut RED_FilmBias_Rec2020_N-Log_to_Rec709_BT1886.cube out.jpg
convert: image sequence is required `-clut' @ error/mogrify.c/MogrifyImageList/7890.
convert: no decode delegate for this image format `' @ error/constitute.c/ReadImage/746.
convert: no images defined `out.jpg' @ error/convert.c/ConvertImageCommand/3362.

**Comment by fmw42** [comment]
Try

`magick cube:LUT.cube[8] HALD8.png`

to convert the cube LUT to an 8 level HALD LUT image. Then

`magick some_image.png HALD8.png -hald-clut result_image.png`

Use magick in place of convert on IM 7.  See also https://usage.imagemagick.org/color_mods/#hald-clut

**Comment by hadmut** [comment]
I had seen hints in the web, but it didn't work under Ubuntu. (It works under Alpine.)

1. What is the difference between the magick command (which does not exist in the Ubuntu/Debian package) and the convert command?
2. What is the -clut option good for then? 
3. Isn't it a loss of quality if the detailed cube file has to be converted to a png with rather limited resolution, thus introducing rounding errors?

**Reply by fmw42** [reply]
1) magick is convert in IM 7, but much enhanced. convert is for IM 6 but does not have all the features in IM 7 and probably not the cube LUT conversion

2) -clut is for normal LUTs.  hald-clut is for use with hald LUTs.  Hald LUTS can deal with more than 8-bits per channel.

3) No, the quality depends upon how many levels are in the HALD LUT

**Comment by hadmut** [comment]
What is "normal LUTs", if not .cube files? Cube files are the most common, or at least one of the most common LUT formats.

**Reply by fmw42** [reply]
Please read https://usage.imagemagick.org/color_mods/#color_lut and the subsequent sections on cluts. 

Please also see:

http://www.imagemagick.org/discourse-server/viewtopic.php?f=1&t=9620
http://www.imagemagick.org/script/command-line-processing.php
https://imagemagick.org/script/command-line-options.php (and see the sitemap link at the bottom)
http://www.imagemagick.org/Usage/reference.html
http://www.imagemagick.org/Usage/
https://github.com/ImageMagick/usage-markdown
https://imagemagick.org/script/porting.php#cli
https://legacy.imagemagick.org/discourse-server/index.php
